haskell/test/SMT/Common/Run.hs

```diff
@@ -0,0 +1,254 @@
+-- SPDX-FileCopyrightText: 2021 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+module SMT.Common.Run
+  ( runBaseDaoSMT
+  ) where
+
+import Universum hiding (drop, swap)
+
+import qualified Data.Map as Map
+import Data.Sequence (Seq(..))
+import Fmt (pretty)
+import Hedgehog
+
+import Lorentz hiding (now, (>>))
+import qualified Lorentz.Contracts.Spec.FA2Interface as FA2
+import Lorentz.Test.Consumer (contractConsumer)
+import Michelson.Untyped.Entrypoints (unsafeBuildEpName)
+import Morley.Nettest
+import qualified Morley.Micheline as MM
+import Morley.Nettest.Pure
+
+import Ligo.BaseDAO.Contract
+import Ligo.BaseDAO.Types
+import SMT.Model.BaseDAO.Contract
+import SMT.Model.BaseDAO.Types
+import SMT.Common.Gen
+import SMT.Common.Types
+import Test.Ligo.BaseDAO.Common.StorageHelper as StorageHelper
+import Test.Ligo.BaseDAO.Common
+
+-- | The functions run the generator to get a list of entrypoints
+-- and do setups before calling entrypoints againts ligo and haskell model
+-- For Ligo:
+--   - Setup initial level (generated by the generator)
+--   - Originate auxiliary contracts
+--   - Originate basedao contract
+-- For Haskell:
+--   - Setup `ModelState`
+runBaseDaoSMT :: SmtOption -> PropertyT IO ()
+runBaseDaoSMT option@SmtOption{..} = do
+
+  -- Run the generator to get a function that will generate a list of entrypoint calls.
+  mkModelInput <- forAll (runGeneratorT (genMkModelInput option) $ initGeneratorState soMkPropose)
```

Comment:
Have you used hedgehog's `collect`/`label` to see if the distribution of our generators is good enough?

E.g., if a property-based test generates 10 actions on average, and 90% of the time the emulator fails on the 2nd action and ignores the other 8, then the tests are really not that helpful. In that case, we'd need to find out why they're failing early so often, and adjust the distribution of the generators.

This issue (unbalanced generators) is extremely common in complex property-based tests like this. So we should collect some statistics using `collect`/`label` to rule that out.

Some questions that could give us some insight into this:
1. How many actions does a test generate on average?
2. How many of the generated actions are successfully interpreted? I would expect this percentage to be relatively high.

Reply:
Ok, I missed something: when we fail to interpret an action, we don't stop there, we move onto the next action, and interpret _all_ of them. So the example I described above was not really accurate:

> E.g., if a property-based test generates 10 actions on average, and 90% of the time the emulator fails on the 2nd action and ignores the other 8, then the tests are really not that helpful. In that case, we'd need to find out why they're failing early so often, and adjust the distribution of the generators.

Still, if 90% of the generated actions fail, then the tests are not really meaningful, because failing actions don't move the contract's internal state forward. E.g., if all actions fail, then the contract's final state will be the exact same as its initial state, so the test is less meaningful.

So the questions I asked above still stand.

Reply:
> 1. How many actions does a test generate on average?
It generates between 8 - 12 calls per round (Each round uses the same storage), and we run the test 30 rounds.

> 2. How many of the generated actions are successfully interpreted? 
Pretty much all of them, since the tests passed. 

The problem is that some of the calls result in failure cases. Ideally, we want to test the successful cases as much as the failure cases.

As a result, I create a `proposing` group calls (each calls are more likely to result in successful cases), and another group contains completely random calls. And we randomly select one of them per round.

> Have you used hedgehog's collect/label to see if the distribution of our generators is good enough?

Is it possible to use this to detect the how many successful/failure cases are reached? Currently, I'm just doing this manually to create a group of call that results successful cases.